### PR TITLE
ArC: add InterceptionProxySubclass

### LINKED ISSUE #44949: Allow for type-safe delegate handle with intercepted beans with just specified interface, but no exact bean class
### Description

When you define a synthetic bean like

```
                var itemBuilder = SyntheticBeanBuildItem
                        .configure(CHAT_MODEL) // <<-- interface class
                        .setRuntimeInit()
                        .defaultBean()
                        .scope(ApplicationScoped.class)
                        .createWith(recorder.chatModel(config, configName))
                        .injectInterceptionProxy(ResponseInterceptorBindingSource.class);

...

            return new Function<>() {
                @Override
                public ChatLanguageModel apply(SyntheticCreationalContext<ChatLanguageModel> context) {
                    InterceptionProxy<ChatLanguageModel> proxy = context.getInterceptionProxy();
                    return proxy.create(builder.build());
                }

...

@ResponseInterceptorBinding
public abstract class ResponseInterceptorBindingSource {
}
```

you get back - apart from ClientProxy - also a generated class,
where the only way to get to the actual `delegate` is currently via reflection:

```
    public static <T> T unwrap(T target) {
        T sub = ClientProxy.unwrap(target);
        if (sub instanceof Subclass) {
            try {
                Field delegate = sub.getClass().getDeclaredField("delegate");
                delegate.setAccessible(true);
                sub = (T) delegate.get(sub);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
        return sub;
    }
```

### Implementation ideas

Something similar to `ClientProxy` would be handy.

### PR DESCRIPTION
**Pull request by Ladicek**
Fixes #44949.

### CODE REVIEW

**Comment by Ladicek** [comment]
I'm not exactly sure about the name `InterceptionProxySubclass`. If we ever add `DecorationProxy`, that proxy will likely implement this interface as well, and the name will be misleading. But that's a premature concern, the decoration proxy might end up implementing another interface as well, and if not, there are ways how to change the name in a compatible fashion. So not a big deal.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ClientProxy.java [hunk 1]
```diff
@@ -42,14 +42,14 @@ public interface ClientProxy {
...
-     * @param obj
+     * @param <T> the type of the object to unwrap
+     * @param obj the object to unwrap
      * @return the contextual instance if the object represents a client proxy, the object otherwise
      */
     @SuppressWarnings("unchecked")
     static <T> T unwrap(T obj) {
-        if (obj instanceof ClientProxy) {
-            return (T) ((ClientProxy) obj).arc_contextualInstance();
+        if (obj instanceof ClientProxy proxy) {
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ClientProxy.java:51]: :+1: I really like pattern matching for instanceof!

**Comment by Ladicek** [comment]
I honestly have no idea what

> Failed to get latest Mandrel release for Java 21 from DISCO API: TypeError: Cannot read properties of undefined (reading 'links')

means.

**Review by manovotn — APPROVED** [review]

**Comment by Ladicek** [comment]
Rebased and force-pushed. Hopefully the previous error was just a transient failure and CI will get through this time.

**Comment by gsmet** [comment]
I think the previous error was about the API used to resolve the versions of Mandrel not being available so no biggie.

The new one is a `io.smallrye.mutiny.TimeoutException`, I think we can merge.

**Pull request merged**